cpuid/cpuid.go

```diff
@@ -0,0 +1,26 @@
+package cpuid
+
+import (
+	"github.com/bobuhiro11/gokvm/kvm"
+)
+
+func cpuid_low(arg1, arg2 uint32) (eax, ebx, ecx, edx uint32) // implemented in cpuid.s
+
+func CPUID(leaf uint32) (uint32, uint32, uint32, uint32) {
+	return cpuid_low(leaf, 0)
+}
+
+// patchCPUID patches CPUIDs before vcpu generation.
+func PatchCPUID(ids *kvm.CPUID, patches []kvm.CPUIDEntry2) {
+	for _, id := range ids.Entries {
+		for _, patch := range patches {
+			if id.Function == patch.Function && id.Index == patch.Index {
+				id.Flags |= 1 << patch.Flags
+				id.Eax |= 1 << patch.Eax
```

Comment:
That is correct, but....

I think one patch should only manipulates one bit in one specific CPUIDEntry2. We can argue about if using the CPUIDEntry2 structure is the optimal to use in both cases.

I would argue it is, because it does not create addional overhead with a seperate structure.
Is it the "clean" solution? Well, I would say no, because we use one structure type in two seperate ways.

What do you think?

Reply:
Hmmm, I would like to avoid using the same data structure kvm.CPUIDEntry2 in two different senses, bit position and bit sequence, as it is a source of confusion.

If we want to restrict one patch to operate on just one bit, how about explicit validation?

```go
                if bits.OnesCount32(patch.Eax)+bits.OnesCount32(patch.Ebx)+
                        bits.OnesCount32(patch.Ecx)+bits.OnesCount32(patch.Edx)+
                        bits.OnesCount32(patch.Flags) != 1 {
                        return errors.New("One patch should contain single bit")
                }
```

Reply:
Done.